crmsh/ui_corosync.py

```diff
@@ -47,6 +50,221 @@ def _diff_nodes(args):
+            if nodeid == -1:
+                logger.error(f'Unknown node {name}.')
+            node_addresses[nodeid] = addr
+        try:
+            self.__save_changed_config(
+                lm,
+                lm.add_link(node_addresses, args.options),
+                reload=True,
+            )
+        except corosync.LinkManager.MissingNodesException as e:
+            for nodeid in e.nodeids:
+                name = next(x.name for x in nodes if x.nodeid == nodeid)
+                logger.error('The address of node %s is not specified.', name)
+            logger.error('The address of all nodes must be specified when adding a new link.')
+            return False
+        except corosync.LinkManager.DuplicatedNodeAddressException as e:
+            node1 = next(x.name for x in nodes if x.nodeid == e.node1)
+            node2 = next(x.name for x in nodes if x.nodeid == e.node2)
+            logger.error('Duplicated ip address %s=%s: already used for %s=%s.', e.node1, e.address, e.node2, e.address)
+            return False
+
+
+    @command.completer(completers.call(lambda: [
+        str(link.linknumber)
+        for link in corosync.LinkManager.load_config_file().links()
+        if link.linknumber != 0
+    ]))
+    def do_remove(self, context, linknumber: str):
+        if not linknumber.isdecimal():
+            raise ValueError(f'Invalid linknumber: {linknumber}')
+        linknumber = int(linknumber)
+        lm = self.__load_and_validate_links()
+        self.__save_changed_config(
+            lm,
+            lm.remove_link(linknumber),
+            reload=True,
+        )
+
+    @staticmethod
+    def _validate_node_addresses(node_addrs: typing.Mapping[str, str]):
+        node_interfaces = {
+            node: iproute2.IPAddr(json.loads(stdout)).interfaces()
+            for node, (_, stdout, _) in parallax.parallax_call(node_addrs.keys(), 'ip -j addr')
+        }
+        for node, addr in node_addrs.items():
+            ip_addr = ipaddress.ip_address(addr)
+            if not any(
+                ip_addr == addr.ip
+                for interface in node_interfaces[node]
+                for addr in interface.addr_info
+            ):
+                raise ValueError(f'{addr} is not a configured interface address on node {node}.')
+
+    @staticmethod
+    def __load_and_validate_links():
+        lm = corosync.LinkManager.load_config_file()
+        if lm.totem_transport() != 'knet':
+            raise ValueError('Corosync is not using knet transport')
+        return lm
+
+    @staticmethod
+    def __save_changed_config(linkmanager, dom, reload: bool):
+        linkmanager.write_config_file(dom)
+        nodes = utils.list_cluster_nodes()
+        nodes.remove(utils.this_node())
+        logger.info('Synchronizing corosync.conf in the cluster...')
+        if not corosync.push_configuration(nodes):
+            raise ValueError('Failed to synchronize corosync.conf in the cluster.')
+        if reload:
```

Comment:
It's great `crm corosync link` can work without the running cluster.
However, there is no point to run corosync-cfgtool -R in this case.

```suggestion

        if not ServiceManager(sh.ClusterShellAdaptorForLocalShell(sh.LocalShell())).service_is_active("corosync.service"):
            return

        if reload:
```